src/components/views/rooms/DecryptionFailureBar.tsx

```diff
@@ -0,0 +1,208 @@
+    useEffect(() => {
+        const timeout = setTimeout(() => setWaiting(false), WAIT_PERIOD);
+        return () => clearTimeout(timeout);
+    }, []);
+
+    const [needsVerification, setNeedsVerification] = useState<boolean>(false);
+    const [hasOtherVerifiedDevices, setHasOtherVerifiedDevices] = useState<boolean>(false);
+    const [hasKeyBackup, setHasKeyBackup] = useState<boolean>(false);
+
+    const [requestedSessions, setRequestedSessions] = useState<Set<string>>(new Set());
+
+    useEffect(() => {
+        if (needsVerification || !hasOtherVerifiedDevices) return;
+
+        const newRequestedSessions = new Set<string>();
+        for (const event of failures) {
+            const sessionId = event.getWireContent().session_id;
+            if (!sessionId || requestedSessions.has(sessionId) || newRequestedSessions.has(sessionId)) continue;
+            newRequestedSessions.add(sessionId);
+            context.cancelAndResendEventRoomKeyRequest(event);
+        }
+        if (newRequestedSessions.size > 0) {
+            setRequestedSessions(oldRequestedSessions => {
+                const requestedSessions = new Set(oldRequestedSessions);
+                for (const session of newRequestedSessions) {
+                    requestedSessions.add(session);
+                }
+                return requestedSessions;
+            });
+        }
+    }, [needsVerification, hasOtherVerifiedDevices, failures, room, requestedSessions, context]);
+
+    const updateDeviceInfo = useCallback(async () => {
+        context.crypto.deviceList.invalidateUserDeviceList(context.getUserId());
+        await context.crypto.deviceList.refreshOutdatedDeviceLists();
+
+        const deviceId = context.getDeviceId();
+        let verified = true; // if we can't get a clear answer, don't bug the user about verifying
+        try {
+            verified = context.checkIfOwnDeviceCrossSigned(deviceId);
+        } catch (e) {
+            console.error("Error getting device cross-signing info", e);
+        }
+        setNeedsVerification(!verified);
+
+        let otherVerifiedDevices = false;
+        try {
+            const { devices } = await context.getDevices();
+            otherVerifiedDevices = devices.some(
+                (device) => device.device_id !== deviceId && context.checkIfOwnDeviceCrossSigned(device.device_id),
+            );
+        } catch (e) {
+            console.error("Error getting info about other devices", e);
+        }
+        setHasOtherVerifiedDevices(otherVerifiedDevices);
+
+        let keyBackup = false;
+        try {
+            const keys = await context.isSecretStored('m.cross_signing.master');
+            keyBackup = (keys !== null && Object.keys(keys).length > 0);
+        } catch (e) {
+            console.error("Error getting info about key backups", e);
+        }
+        setHasKeyBackup(keyBackup);
+    }, [context]);
+
+    useEffect(() => {
+        updateDeviceInfo().catch(console.error);
```

Comment:
Is this necessary or the best we can do here? It looks like a bit of surprising thing to have in production code, but I'm out of touch here

Reply:
This update isn't fundamentally crucial to the operation of this component, and if any of the methods we're calling fail, it shouldn't crash the DecryptionFailureBar as a whole. If `updateDeviceInfo` never successfully runs, then we'll just be displaying the generic "requesting keys" message, since we can't give any more concrete call to action.